src/helpers/buffers.rs

```diff
@@ -0,0 +1,139 @@
+enum ReceiveBufItem {
+    /// There is an outstanding request to receive the message but this helper hasn't seen it yet
+    Requested(oneshot::Sender<Box<[u8]>>),
+    /// Message has been received but nobody requested it yet
+    Received(Box<[u8]>),
+}
+
+impl SendBuffer {
+    pub fn new(max_channel_capacity: u32) -> Self {
+        Self {
+            max_capacity: max_channel_capacity as usize,
+            inner: HashMap::default(),
+        }
+    }
+
+    pub fn push(
+        &mut self,
+        channel_id: ChannelId,
+        msg: MessageEnvelope,
+    ) -> Option<Vec<MessageEnvelope>> {
+        let vec = match self.inner.entry(channel_id) {
+            Entry::Occupied(entry) => {
+                let vec = entry.into_mut();
+                vec.push(msg);
+
+                vec
+            }
+            Entry::Vacant(entry) => {
+                let vec = entry.insert(Vec::with_capacity(self.max_capacity));
+                vec.push(msg);
+
+                vec
+            }
+        };
+
+        if vec.len() >= self.max_capacity {
+            let data = mem::replace(vec, Vec::with_capacity(self.max_capacity));
+            Some(data)
+        } else {
+            None
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.inner.len()
+    }
+
+    pub fn remove_random(&mut self) -> (ChannelId, Vec<MessageEnvelope>) {
+        assert!(self.len() > 0);
+
+        let channel_id = self.inner.keys().next().unwrap().clone();
+        let data = self.inner.remove(&channel_id).unwrap();
+
+        (channel_id, data)
+    }
+}
+
+impl ReceiveBuffer {
+    pub fn new() -> Self {
+        Self {
+            inner: HashMap::default(),
+        }
+    }
+
+    /// Process request to receive a message with the given `RecordId`.
+    pub fn receive_request(
+        &mut self,
+        channel_id: ChannelId,
+        record_id: RecordId,
+        sender: oneshot::Sender<Box<[u8]>>,
+    ) {
+        match self.inner.entry(channel_id).or_default().entry(record_id) {
+            Entry::Occupied(entry) => match entry.remove() {
+                ReceiveBufItem::Requested(_) => {
+                    panic!("More than one request to receive a message for {record_id:?}");
+                }
+                ReceiveBufItem::Received(payload) => {
+                    sender.send(payload).unwrap_or_else(|_| {
+                        tracing::warn!("No listener for message {record_id:?}");
+                    });
+                }
+            },
+            Entry::Vacant(entry) => {
+                entry.insert(ReceiveBufItem::Requested(sender));
+            }
+        }
+    }
+
+    /// Process message that has been received
+    pub fn receive_messages(&mut self, channel_id: &ChannelId, messages: Vec<MessageEnvelope>) {
+        for msg in messages {
+            match self
+                .inner
+                .entry(channel_id.clone())
+                .or_default()
+                .entry(msg.record_id)
```

Comment:
This bothers me a little.  We might want to consider adding new structs rather than having this one work its way down through two hashmaps (do this later).

Reply:
same, I don't like nested maps as well, but hopefully we can get rid of the second map and replace it with a vec indexed by record_id + offset